packages/theia-integration/src/node/websocket-connection-forwarder.ts

```diff
@@ -0,0 +1,73 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { WebSocketMessageReader, WebSocketMessageWriter, WebSocketWrapper } from '@eclipse-glsp/protocol';
+import { Channel, Disposable, DisposableCollection, MessageProvider } from '@theia/core';
+import { createMessageConnection } from '@theia/core/shared/vscode-languageserver-protocol';
+import { Message } from 'vscode-jsonrpc/node';
+import { WebSocket } from 'ws';
+/**
+ * Creates a new {@link MessageConnection} on top of a given WebSocket and forwards messages from service channel to this connection
+ */
+export class WebSocketConnectionForwarder implements Disposable {
+    protected toDispose = new DisposableCollection();
+
+    constructor(protected readonly channel: Channel, protected readonly webSocket: WebSocket) {
+        const wrappedWebSocket = wrapWebSocket(webSocket);
+        const reader = new WebSocketMessageReader(wrappedWebSocket);
```

Comment:
I think we  could simplify the forwarding here. It should not be necessary to install reader/writers on top of the websocket. IMO we should be able to simply directly forward the message buffer received from the channel and vice versa.